python/sources/simple-csv-new/connector/source_connector.py

```diff
@@ -0,0 +1,178 @@
+import csv
+import logging
+import json
+from typing import Optional, Union, Callable, Any, Iterator
+
+from .connector_templates import (
+    SourceConsumer,
+    SourceProducer,
+    SourceConnector,
+    setup_logging
+)
+from quixstreams.logging import LogLevel
+from quixstreams.models.messages import KafkaMessage
+from quixstreams.models.serializers import SerializerType
+from quixstreams.models.types import MessageKey, MessageValue
+
+
+logger = logging.getLogger(__name__)
+
+__all__ = ("CsvSourceConnector",)
+
+
+class CsvConsumer(SourceConsumer):
+    """
+    A thin wrapper around csv.DictReader for lazy row-based reading of a CSV file
+    with a header.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        key_parser: Optional[Union[str, Callable[[dict], MessageKey]]] = None,
+        value_parser: Optional[Callable[[dict], MessageValue]] = json.dumps,
+        **csv_reader_kwargs,
+    ):
+        """
+        :param path: file path to a .csv file
+        :param key_parser: how a key is generated for a given data row. Options:
+            - None; no key will be generated (default)
+            - A string, representing a column name; uses that row's column value.
+            - A function that handles the row as a dict.
+        :param value_parser: A function for generating resulting message value for the
+            row. Default: json.dumps
+        :param csv_reader_kwargs: kwargs passthrough to csv.DictReader
+        """
+        self._path = path
+        self._key_parser = key_parser
+        self._value_parser = value_parser
+        self._file = None
+        self._reader = None
+        self._reader_kwargs = csv_reader_kwargs
+
+    def _as_message(self, data: dict) -> KafkaMessage:
```

Comment:
I think this can be virtual method in base class, what is specific about this implementation to CSV?

Reply:
I didn't make this a virtual method as it's intention/requirement is already captured by the other virtual `consume` method; I just did this as a bit of sugar/make the app nicer to read. It could be removed entirely and placed inside the consume method, really